---
title: "Clustering cells and nonlinear dimensional reduction"
authors: "Karla Lindquist (with Angelo Pelonero and Rebecca Jaszczak)"
date: "3/30/2021"
output: html_notebook
---

</br>

#### Cluster the cells 

Seurat v3 applies a graph-based clustering approach, building upon initial strategies in (Macosko et al). Importantly, the distance metric which drives the clustering analysis (based on previously identified PCs) remains the same. However, the v3 approach (compared to v2) to partioning the cellular distance matrix into clusters has dramatically improved.

The new(er) approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data [SNN-Cliq, Xu and Su, Bioinformatics, 2015](http://bioinformatics.oxfordjournals.org/content/early/2015/02/10/bioinformatics.btv088.abstract) and CyTOF data [PhenoGraph, Levine et al., Cell, 2015](http://www.ncbi.nlm.nih.gov/pubmed/26095251).

Briefly, these methods embed cells in a graph structure - for example a `K-nearest neighbor (KNN) graph`, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected ‘quasi-cliques’ or ‘communities’.

As in `PhenoGraph`, Seurat v3 first constructs a KNN graph based on the euclidean distance in PCA space, and refines the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the `FindNeighbors` function, and takes as input the previously defined dimensionality of the dataset (first 10 PCs).

To cluster the cells, Seurat v3 next applies modularity optimization techniques such as the `Louvain algorithm` (default) or `SLM` [SLM, Blondel et al., Journal of Statistical Mechanics](http://dx.doi.org/10.1088/1742-5468/2008/10/P10008), to iteratively group cells together, with the goal of optimizing the standard modularity function.

The `FindClusters` function implements this procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between `0.4-1.2` typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the `Idents` function.  

```{r}
# TRY: sequence of resolutions (.4-2, plot each, pick favorite)
# pay attention to your dims - keep them consistent in this notebook
pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)
```

```{r}
# Look at cluster IDs of the first 5 cells
head(Idents(pbmc), 5)
```

</br> 

#### Run non-linear dimensional reduction (UMAP/tSNE)

Seurat offers several non-linear dimensional reduction techniques, such as `tSNE` and `UMAP`, to visualize and explore these datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the `UMAP` and `tSNE`, we STRONGLY suggest using the same PCs as input to the clustering analysis.  

```{r}
# UMAP now runs in entirely in R (it used to require Python and the reticulate package)
pbmc <- RunUMAP(pbmc, dims = 1:10)
# pbmc <- RunTSNE(pbmc, dims = 1:10)
```

What is the difference between UMAP and tSNE? Which one is better? This is beyond the scope of this workshop but there is a good article here called ["tSNE vs. UMAP: Global Structure"](https://towardsdatascience.com/tsne-vs-umap-global-structure-4d8045acba17) by Nikolay Oskolkov (towardsdatascience.com, 2020).  


Again, you can save the output to read in later - the steps above should not take too long to run so we ran them, but in case something did not work for you or you got stuck, let's read in the output.  
```{r}
projdir <- sub("src", "", getwd()) 

# saveRDS(pbmc, file = paste0(projdir, "./output/pbmc_tutorial_umap.rds"))
# readRDS(paste0(projdir, "./output/pbmc_tutorial_umap.rds"))
```


</br> 

Now we will use `DimPlot()` function again. Note that you can set the label argument to TRUE, or use the `LabelClusters()` function to help label individual clusters. If the reduction = "umap" argument is giving you trouble, try "tsne".   

```{r}
DimPlot(pbmc, reduction = "umap")
```

Take a close look at the various arguments available, such as group.by and cells.highlight.  

```{r}
?DimPlot
```


</br>

#### Finding differentially expressed features (cluster biomarkers)

Seurat can help you find markers that define clusters via differential expression. By default, it identifes positive and negative markers of a single cluster (specified in `ident.1`), compared to all other cells. `FindAllMarkers` automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.

The `min.pct` argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a feature to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of features that are unlikely to be highly discriminatory. As another option to speed up these computations, `max.cells.per.ident` can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significiant and the most highly differentially expressed features will likely still rise to the top.  


```{r}
# find all markers of cluster 1
cluster1.markers <- FindMarkers(pbmc, ident.1 = 1, min.pct = 0.25, verbose=FALSE)
head(cluster1.markers, n = 5)
```

```{r}
# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25, verbose=FALSE)
head(cluster5.markers, n = 5)
```

```{r}
# find markers for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25, verbose=FALSE)
pbmc.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
pbmc.markers.0 = pbmc.markers[pbmc.markers$cluster == 0,] #example of subsetting the FindAllMarkers result for Cluster 0
```


Save or read in the results.    
```{r}
projdir <- sub("src", "", getwd()) 

# saveRDS(pbmc.markers, file = paste0(projdir, "./output/pbmc_tutorial_markers.rds"))
# readRDS(paste0(projdir, "./output/pbmc_tutorial_markers.rds"))
```


Seurat has several tests for differential expression which can be set with the test.use parameter (see the DE vignette for details). For example, the ROC test returns the ‘classification power’ for any individual marker (ranging from 0 = random, to 1 = perfect).

```{r}
cluster1.markers <- FindMarkers(pbmc, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE, verbose=FALSE)
```

Seurat includes several tools for visualizing marker expression. The most commonly used visualizations are `VlnPlot()` to show expression probability distributions across clusters, and `FeaturePlot()` to show feature expression on a tSNE or PCA plot. We also suggest exploring `RidgePlot()`, `CellScatter()`, and `DotPlot()` as additional methods to view this dataset.  
```{r}
VlnPlot(pbmc, features = c("NKG7", "CD79A"))
```

```{r}
# you can plot raw counts as well
VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = "counts", log = TRUE)
```

```{r}
FeaturePlot(pbmc, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", 
    "CD8A"))
```


`DoHeatmap()` generates an expression heatmap for given cells and features. In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.

```{r}
top10 <- pbmc.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```
</br>

#### Assigning cell type identity to clusters 

Fortunately in the case of this dataset, we can use canonical markers to easily match the unbiased clustering to known cell types:
```{r}
id.gene.list = c('IL7R', 'CCR7',		#Naive CD4+ T
                 'IL7R', 'S100A4',	#Memory CD4+
                 'CD14', 'LYZ',	  	#CD14+ Mono
                 'MS4A1',	    		  #B
                 'CD8A',	          #CD8+ T
                 'FCGR3A', 'MS4A7',	#FCGR3A+ Mono
                 'GNLY', 'NKG7',		#NK
                 'FCER1A', 'CST3',	#DC
                 'PPBP')			      #Platelet

FeaturePlot(pbmc, features = id.gene.list)

new.cluster.ids <- c("Naive CD4 T",  #0
                     "CD14+ Mono",   #1
                     "Memory CD4 T", #2
                     "CD8 T",        #3
                     "B",            #4
                     "NK",           #5
                     "FCGR3A+ Mono", #6
                     "DC",           #7
                     "Platelet")     #8
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

Again you can save and read in the results.    
```{r}
projdir <- sub("src", "", getwd()) 

# saveRDS(pbmc, file = paste0(projdir, "./output/pbmc_tutorial_final.rds"))
# readRDS(paste0(projdir, "./output/pbmc_tutorial_final.rds"))
```

</br> 

#### Summary 

In this notebook, we clustered cells into distinct identities based on their differential gene expression. We observed these distinct transcriptional states with PCA, tSNE and/or UMAP. We also explored the different genes making up these cluster identities by playing with the `FindMarkers()` and `FindAllMarkers()` commands. Finally, we were able to use expression data to give meaningful labels to our known cell types.  